sound/soc/sof/intel/hda-ipc.c

```diff
@@ -274,6 +276,27 @@ void hda_ipc_msg_data(struct snd_sof_dev *sdev,
 	}
 }
 
+void hda_ipc_receive_msg(struct snd_sof_dev *sdev, u32 msg_cmd)
+{
+	struct sof_intel_hda_dev *hdev = sdev->pdata->hw_pdata;
+	struct sof_ipc_comp_event event;
+	u32 cmd;
+
+	cmd = msg_cmd & SOF_GLB_TYPE_MASK;
+
+	switch (cmd) {
+	case SOF_IPC_GLB_COMP_MSG:
+		snd_sof_ipc_msg_data(sdev, NULL, &event, sizeof(event));
+
+		/* reschedule D0i3 work if keyword detected */
+		if (event.event_type == SOF_CTRL_EVENT_KD)
+			mod_delayed_work(system_wq, &hdev->d0i3_work,
+					 msecs_to_jiffies(SOF_HDA_D0I3_WORK_DELAY_MS));
```

Comment:
The premise for D0ix is that there will be no IPC between host and DSP. so if the DSP initiates *ANY* type of IPC, shouldn't this also reset the idleness counter? Why should this be done only with Keyword detection?

Reply:
hmm the premise is there are no IPC from the host to the DSP but not for the other direction. Thats what I'm tryng to fix here. But I'm not sure we should do this unconditionally because we also have IPC RX for things like trace.

Reply:
All IPCs are two-way: notification and ack. If there is a trace IPC ack going from the host to the DSP, IMHO this breaks the contract that the host will not interact with the DSP...

Reply:
right but the ack is different from a new message. For that matter even if trace werent enabled, the ACK for the comp notification after the keyword is detected should reset the counter. It would be simpler to do but I want to make sure this is OK

Reply:
Hmm, having special logic for just one type of message does not look right.

How about reseting the D0i3 timer if any FW initiated IPC is received by host? Granted keyword-notification and traces are the only ones we have now, but this would still make more sense and be more future-proof.

This

Reply:
axk @kv2019i . let me switch to resetting the timer when seding the IPC ack as well.